### ISSUE TITLE
CFG and pretrained weight explanation

### ISSUE THREAD

**Issue by raudipra** [pjreddie/darknet#1317]
Hi I am new with this Darknet stuff and I am curious how pretrained weight fitting itself in particular cfg on training mode ( `darknet detector train` ). Let's say in normal situation you would use `yolov3.weight` for `yolov3.cfg` as the pretrained weight. 

But what will happen if those combination doesn't match? 

For example, I created my own cfg file which has very different layer configuration compared to yolov3.cfg, I name it random.cfg. What will happen if I run command like this:
`darknet detector train cfg/coco.data cfg/random.cfg yolov3.weight`

Will it train the network from scratch? or it will force the cfg so it will fit the weight reference? or another combination technique that I can't think of?

**Comment by PeterQuinn925** [comment]
The short answer is that the weights file is updated based on the CFG file not the other way around.

If you modify the network by editing the CFG and train, it will create a new weights file starting with the one you put in the command line. If you are starting with a new dataset and a new cfg, you should leave the weight file blank and it will start training with random values. It puts a copy of the new weights file in the backup directory after every 100 iterations of training.0

**Comment by raudipra** [comment]
@PeterQuinn925 thank you for answering, I still don't get the "it will create a new weights file starting with the one you put in the command line". Let's put a real example here, here is a snippet of the first part of my original CFG.
`[convolutional]
batch_normalize=1
filters=16
size=3
stride=2
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=8
size=1
stride=1
pad=1
activation=leaky
`
this CFG will have a pretrained weight which consist of 3x3x16+bias weight (from first conv layer) and 1x1x8+bias weight (from second conv layer).

what will happen if I change my CFG file into this (I just change the second conv layer) and using previous pretrained weight on it?
`[convolutional]
batch_normalize=1
filters=16
size=3
stride=2
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=16
size=3
stride=1
pad=1
activation=leaky
`
as you can see the number of weight in second layer (3x3x16) will be different with the pretrained weight (1x1x8), what my point is if from the start the pretrained weight is not match with the configuration how can we train it in our network? Not to mention if I add or delete a layer, the weight configuration will be more different. How does the pretrained weight match this changes of configuration/CFG?

**Comment by PeterQuinn925** [comment]
I don't know what will happen if your pretrained weights don't match the network in the CFG file. I suspect it will seg fault. As a practical answer, if you change the network definition in the CFG, don't supply a pre-trained weight file - it will create a new one as you train.

**Comment by raudipra** [comment]
@PeterQuinn925 at the beginning, I was also expecting that kind of outcome. But based on my experiment, it turns out the training still working without any error that is why I've opened this issue. Thank you anyway

**Comment by naserpiltan** [comment]
@raudipra  I have same problem.did you get to good result??did your network worke fine?
if not ، what we can to do ??Thanks
